### PR TITLE
[WIP] [DONOTMERGE] Bert Featurizer

### PR DESCRIPTION
**Pull request by alat-rights**
# BERT Featurizer

## Adding Support for HuggingFace BERT Tokenizer in DeepChem

This PR replaces #2540 and introduces a wrapper class for the BERT Tokenizer based on (and inheriting from) the Roberta Featurizer in #2523 and #2581 .

**DO NOT MERGE until** 
- mypy errors cleared
- The BertFeaturizer documentation is revised to explain what BERT tokenization does, in line with Bharath's comments
- ~~RobertaFeaturizer is in DeepChem. This class depends on RobertaFeaturizer.~~\*
- ~~run_bert is removed from this PR. run_bert is used for development purposes and is not intended to be added to DeepChem~~ \*\*

\* This class no longer depends on RobertaFeaturizer
\*\* Removed

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/bert_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# Current approach to handling both molecules and strings is a bit messy.
+try:
+  from deepchem.utils.typing import RDKitMol
+except:
+  warnings.warn("""RDKitMol was not found in your environment, so this class
+                will only featurize strings and will not featurize RDKitMol
+                objects.""")
+                
+
+class BertFeaturizer(RobertaFeaturizer, BertTokenizerFast):
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:22]: Isn't BERT featurization less specialized than roberta featurization? Perhaps we should have the inheritance structure be reversed here

CC @seyonechithrananda

File: deepchem/feat/bert_featurizer.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  """Bert Featurizer.
+
+  The Bert Featurizer is a wrapper class for HuggingFace's BertTokenizerFast.
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:25]: Can you rewrite this section to explain conceptually what the tokenizer/featurizer does? Remember that our users may never have used HuggingFace before so it's better to start from first principles and explain the concept

File: deepchem/feat/bert_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  This class intends to allow users to use the BertTokenizer API while
+  remaining inside the DeepChem ecosystem.
+
+  Usage Example (based on HuggingFace and RostLab ProtBert docs)
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:30]: This should just be
```
Examples
--------
```
Check out the numpy doc style https://numpydoc.readthedocs.io/en/latest/format.html

File: deepchem/feat/bert_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  This class inherits first from RobertaFeaturizer and then from
+  BertTokenizerFast. Its logic relies on super() calls in RobertaFeaturizer
+  being delegated to BertTokenizerFast in accordance with Python's multiple
+  inheritance rules.
+
+  Parts of this class, including its documentation and naming scheme, is based
+  on and/or bear similarities to RobertaFeaturizer for the sake of consistency.
+  """
+
+  def _featurize(self, sequence: Union[str, RDKitMol]) -> List[List[int]]:
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:47]: Similar to the discussion in @seyonechithrananda's PR, let's just make this expect strings? The union type here feels a bit awkward

File: deepchem/feat/bert_featurizer.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    """Tokenizes a datapoint with BertTokenizerFast.
+
+      Parameters
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:50]: The formatting on this docstring is off. You might need to fix manually

File: deepchem/feat/bert_featurizer.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# TODO discuss RDKitMol support (perhaps RobertaFeaturizer should not be a moleule featurizer?
+# Current approach to handling both molecules and strings is a bit messy.
+try:
+  from deepchem.utils.typing import RDKitMol
+except:
+  warnings.warn("""RDKitMol was not found in your environment, so this class
+                will only featurize strings and will not featurize RDKitMol
+                objects.""")
+                
+
```
Comment (rbharath) [inline deepchem/feat/bert_featurizer.py:21]: Make sure to add a unit test! Also make sure to add this new class to the docs so it renders on the docs site

**Review by rbharath — COMMENTED** [review]
Just made a first pass!

File: deepchem/feat/bert_featurizer.py (hunk 1 above)
Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:22]: That makes sense to me!

File: deepchem/feat/bert_featurizer.py (hunk 2 above)
Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:25]: Will do!

File: deepchem/feat/bert_featurizer.py (hunk 4 above)
Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:47]: Hmm, that makes sense!

I do want to maintain some consistency with the existing molecular featurizer workflows that expect RDKit molecules, but is that less important for the time being?

File: deepchem/feat/bert_featurizer.py (hunk 5 above)
Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:50]: Will do!

Is the formatting specified in PEP?

I'll be sure to check with other existing DeepChem programs to make sure it's consistent :)

Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:50]: I think it's fixed!

File: deepchem/feat/bert_featurizer.py (hunk 1 above)
Comment (alat-rights) [reply deepchem/feat/bert_featurizer.py:22]: I just edited my file to no longer rely on RobertaFeaturizer. Could we change RobertaFeaturizer accordingly? @seyonechithrananda

**Comment by alat-rights** [comment]
Thank you for the review! 

All that's left for now is
- Rewrite class documentation to better explain what the tokenizer does
- Make the tests pass!!!!!!!!!!!!!!!!

**Comment by alat-rights** [comment]
Now essentially even with #2581

**Comment by alat-rights** [comment]
All tests passing

**Pull request closed without merging**